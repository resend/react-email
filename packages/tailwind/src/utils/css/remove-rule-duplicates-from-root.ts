import type { Root } from 'postcss';
import { removeIfEmptyRecursively } from './remove-if-empty-recursively';

export const removeRuleDuplicatesFromRoot = (root: Root) => {
  root.walkRules((rule) => {
    root.walkRules(rule.selector, (duplicateRule) => {
      if (duplicateRule === rule) return;

      const parent = duplicateRule.parent;
      duplicateRule.remove();
      if (parent) {
        removeIfEmptyRecursively(parent);
      }
    });
  });
};
