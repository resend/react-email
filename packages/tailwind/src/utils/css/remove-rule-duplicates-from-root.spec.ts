import { generate, parse, type StyleSheet } from 'css-tree';
import { removeRuleDuplicatesFromRoot } from './remove-rule-duplicates-from-root';

test('removeRuleDuplicatesFromRoot', () => {
  const stylesheet = parse(`
    .a { color: red; }
    .b { color: blue; }
    .a { color: green; }
  `);

  removeRuleDuplicatesFromRoot(stylesheet as StyleSheet);

  expect(generate(stylesheet)).toMatchSnapshot();
});
