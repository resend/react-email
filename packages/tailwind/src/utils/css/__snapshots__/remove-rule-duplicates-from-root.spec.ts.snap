// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`removeRuleDuplicatesFromRoot 1`] = `".a{color:red}.b{color:blue}"`;
